### PR TITLE
RFC for reference unification

### PR DESCRIPTION
**Pull request by wycats**
[Rendered][rendered]

[rendered]: https://github.com/emberjs/rfcs/blob/ember-data-reference-unification/active/0000-ember-data-reference-unification.md

### CODE REVIEW

**Comment by igorT** [comment]
Several concerns that were raised in the Slack room:
  1. It seems like people would want to manipulate relationships based on just having the reference, and not having the full object. Is this possible?

  2. It will be easy for references to end up in the dom, see for example:
  ```  
  {{each post.comments as |comment|}}
       {{linkTo 'comment' comment.id}}
  {{/each}}
  ```
  If we want to, how do we stop this? How do we enable this use case?

  3. Lets say you want to filer all the comments whose id is not equal to 1. That CP is going to be annoying/hard to write, because you will need to be filtering on relationshipReferences, but the CP will not be able to observe those, it can only observe `comments.@each.id` or something like that.

**Comment by igorT** [comment]
https://github.com/emberjs/data/issues/2264 seems relevant

**Comment by thomassnielsen** [comment]
I like this a whole lot, and it would solve many of the use cases where we today have unnecessary fetches against the server and/or usage of private APIs. I also agree that the drawbacks seems worth it as described. 

What I'm missing currently is details on how (and if) this will affect current APIs. Especially what Igor describes in point 2, about references not leaking into record arrays (see https://github.com/emberjs/data/issues/2375 for a similar issue). As I see it, the current APIs should maintain only showing loaded records, and references should only show up when using the new reference APIs. Any and all template usage should only show loaded records. I think it would be helpful if this is more explicitly stated in the RFC (if this is indeed the intent).

I also think what Igor describes in point 1 should be implemented if possible without increasing complexity too much. Being able to manipulate relationships based on reference would save a lot of API calls and increase performance in several apps I'm currently working on.

**Comment by sandstrom** [comment]
Looks great! We have a few cases of determining if a reference exist without triggering a server-fetch. This would allows us to move off the private APIs we're currently using.

***

Would it make sense to allow the adapter to sub-class these?

For example, a JSON-API adapter may want to sub-class `HasManyReference` to add a `loadNext` method, utilizing [pagination metadata](http://jsonapi.org/format/#fetching-pagination) to load, say, the next 100 comments for a post into the existing `RecordArray`.

**Comment by wycats** [comment]
> It seems like people would want to manipulate relationships based on just having the reference, and not having the full object. Is this supported behavior? Basically setting a belongsTo to a reference.

The current proposal does not support mutation, but I would be happy to add it if people need it. Perhaps: `reference.update({ ... })`. I'm not sure if it makes sense to support updating everything (like `id`) but maybe it is?

> It will be easy for references to end up in the dom, see for example:

I don't understand how your example puts references in the DOM. Can you give me the JS code that produces this?

> How do we enable this use case?

Can you flesh out the use-case?

> Lets say you want to filer all the comments whose id is not equal to 1. That CP is going to be annoying/hard to write, because you will need to be filtering on relationshipReferences, but the CP will not be able to observe those, it can only observe comments.@each.id or something like that

This proposal doesn't address that case. Do you think it's important to solve it now? If so, I can give it some thought.

**Comment by igorT** [comment]
```
{{each post.commentReferences as |comment|}}
     {{linkTo 'comment' comment.id}}
{{/each}}
```

**Comment by fivetanley** [comment]
More discussion on getting the ids in a template here as well: worth reviewing https://github.com/emberjs/data/issues/3025

cc @chancancode

**Comment by igorT** [comment]
Another use case is to want to bind to `isLoading` or `isReloading` of the relationship

**Comment by wecc** [comment]
Comment from @igorT:
> @tomdale and @wycats are to write up a nice example on how to get IDs and how to use isLoaded

**Comment by jdurand** [comment]
I used to do this the non-public (dirty) way with `province.data.country_id`. I upgraded to ember 1.13.2 and ember-data 1.13.4 and the foreign key ids are not accessible through `model.data` anymore.

I'd like to quick-fix this with a dirty (or not) solution.
Did anyone figure out how to do this without having to patch their API server?

File: active/0000-ember-data-reference-unification.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+```js
+/**
+  If the `remoteType` is `link`, the URL to use to load the relationship.
+
+  @property
+*/
+```
+
+### `ids`
```
Comment (fivetanley) [inline active/0000-ember-data-reference-unification.md]: i think this should be `id` as it is a belongsTo and will only have one id, thus not being plural ids

**Comment by fivetanley** [comment]
I really like the direction this is headed. It seems like a lot of things that were not possible using public APIs are now possible in JavaScript land. I'm wondering what this means in templates though; it seems like it may be a bit cumbersome to get these references yourself every time and define them in a routable component.

API of my dreams:

```handlebars
{{#if comment.isLoaded}}
  your commentID is {{comment.id}}
{{else}}
  <button {{action "load comment"}}>
    Load comment
  </button>
{{/if}}
```

With this current RFC, I don't think this is quite possible without always defining isLoaded, etc yourself in every component or every model.

**Comment by arenoir** [comment]
@jdurand I am also looking for a quick fix to accessing foreign keys.  Did you have any success?

**Comment by jdurand** [comment]
@arenoir no, I haven't had time to dig into this yet.

**Comment by arenoir** [comment]
@jdurand okay here is my "quick fix". Added overrode the normalizeAttributes hook in the serializer to copy the foreign key.  IMO the foreign_key should be public and not be thrown away.

```
//models/checklist-item.js
export default DS.Model.extend({
  timeEntry: belongsTo('time-entry', {async: true}),
  timeEntryId: attr('string'), //set in serializer
});

//serializers/checklist-item.js
export default DS.ActiveModelSerializer.extend({
  normalizeAttributes(typeClass, hash) {
    var foreignKey = hash['timeEntry'];
    
    if (foreignKey) {
      hash['timeEntryId'] = foreignKey;
    }

    return this._super(typeClass, hash);
  }
});
```

I think this a pretty standard use case. Hope this helps.

**Comment by sandstrom** [comment]
This would be a great addition to Ember Data! :boat:

@wycats do you want to wait for more feedback, or is this RFC ready for activation?

**Comment by wycats** [comment]
@sandstrom I'm happy with it, assuming it gets further iterated on over time to handle use-cases we missed in the initial design.

**Comment by sandstrom** [comment]
@wycats Sounds great! I've tried to compile what people have said above (please add if I've missed something).

- [ ] Should mutation be added, or is reading enough for the first version?
- [x] References ending up in the DOM[A]. As I've understood it this won't happen with this API.
- [ ] Use-case: "Filtering all comments whose id is not equal to 1". Should this be supported? Current proposal doesn't cover this.
- [ ] How will `isLoaded` be handled? It was said above that Tom and Yehuda would write up an example. Is such an example still needed?

This would be *very* useful in our app, we're currently going deep into private APIs to work around this[B]. I'd rather see a smaller implementation soon, than wait for the utopian RFC covering all use-cases. :smile: 

Carbon copy: @igorT @wecc @fivetanley 

***

*A. Leaking references into DOM*
```handlebars
{{! will this leak references? I don't think it will, but please correct me if I'm wrong}}
{{each post.commentReferences as |comment|}}
  {{linkTo 'comment' comment.id}}
{{/each}}
```

*B. Brittle Workaround*
```js
// brittle workaround
model._internalModel._relationships.initializedRelationships.branches.canonicalState.length
```

**Comment by sandstrom** [comment]
Friendly ping :smile: @igorT @wecc @fivetanley @wycats 

This is a great proposal (that would solve a concrete pain in our app).

On the checklist above, what are your thoughts? (It's a list of all things mentioned in this thread, I think this should be ready as soon as you guys feel that they are resolved; can add to the list if you feel something is missing)

**Comment by sandstrom** [comment]
Should this be closed now, since https://github.com/emberjs/data/pull/3303 is merged?

**Comment by fivetanley** [comment]
merged as 074c798138fc9d4206c609182614cf3bf24b5fc9. thanks for the reminder @sandstrom!

**Comment by ef4** [comment]
FYI, I just edited the above original post to make "[Rendered](https://github.com/emberjs/rfcs/blob/ce19612e447e0ec393f842bfb31add96edaba9a1/text/0057-ember-data-reference-unification.md)" be a valid permalink. It was broken because it pointed at a file under `active` on `master`.

**Pull request closed without merging**